### PR TITLE
refactor(backend): eslint-plugin-type-graphql + fixes

### LINKED ISSUE #2735: 💥 [DevOps] Eslint for typegraphql
## 💥 DevOps ticket

We need `typegraphql` linting to ensure return types of functions are correctly defined. Otherwise we can get runtime errors. See for discussion and details: https://github.com/gradido/gradido/pull/2706

Implement this:
https://www.npmjs.com/package/eslint-plugin-type-graphql

Especially this: https://github.com/borremosch/eslint-plugin-type-graphql/blob/HEAD/docs/rules/invalid-decorated-type.md

### PR DESCRIPTION
**Pull request by ulfgebhardt**
## 🍰 Pullrequest

This is the same as #2744 but includes all fixes required to have the lint green again

### Issues

- based on #2744
- fixes #2735 

### Todo

- [ ] test with production data locally

### CODE REVIEW

File: backend/src/graphql/model/Balance.ts [hunk 1]
```diff
@@ -19,14 +19,14 @@ export class Balance {
   @Field(() => Decimal)
   balance: Decimal
 
-  @Field(() => Number, { nullable: true })
+  @Field(() => Int, { nullable: true })
```
Comment (Mogge) [inline backend/src/graphql/model/Balance.ts:22]: This is not an integer. Should be Numer

File: backend/src/graphql/model/GdtEntryList.ts [hunk 2]
```diff
@@ -16,15 +19,15 @@ export class GdtEntryList {
...
-  @Field(() => Number)
+  @Field(() => Int)
   count: number
 
   @Field(() => [GdtEntry], { nullable: true })
-  gdtEntries?: GdtEntry[]
+  gdtEntries: GdtEntry[] | null
 
-  @Field(() => Number)
+  @Field(() => Int)
```
Comment (Mogge) [inline backend/src/graphql/model/GdtEntryList.ts:28]: Float?

File: backend/src/graphql/model/GdtEntryList.ts [hunk 3]
```diff
@@ -16,15 +19,15 @@ export class GdtEntryList {
...
   gdtSum: number
 
-  @Field(() => Number)
+  @Field(() => Int)
```
Comment (Mogge) [inline backend/src/graphql/model/GdtEntryList.ts:31]: What is this?

File: backend/src/graphql/resolver/GdtResolver.ts [hunk 4]
```diff
@@ -35,7 +39,7 @@ export class GdtResolver {
   }
 
   @Authorized([RIGHTS.GDT_BALANCE])
-  @Query(() => Number)
+  @Query(() => Int, { nullable: true })
```
Comment (Mogge) [inline backend/src/graphql/resolver/GdtResolver.ts:42]: No Int- Number or Float

File: backend/src/graphql/resolver/TransactionLinkResolver.ts [hunk 5]
```diff
@@ -128,7 +128,7 @@ export class TransactionLinkResolver {
   @Authorized([RIGHTS.QUERY_TRANSACTION_LINK])
   @Query(() => QueryLinkResult)
   async queryTransactionLink(@Arg('code') code: string): Promise<typeof QueryLinkResult> {
-    if (code.match(/^CL-/)) {
+    if (/^CL-/.exec(code)) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionLinkResolver.ts:131]: Why this change?

File: backend/src/graphql/resolver/TransactionLinkResolver.ts [hunk 6]
```diff
@@ -178,7 +178,7 @@ export class TransactionLinkResolver {
     const clientTimezoneOffset = getClientTimezoneOffset(context)
     const user = getUser(context)
 
-    if (code.match(/^CL-/)) {
+    if (/^CL-/.exec(code)) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/TransactionLinkResolver.ts:181]: Same

File: backend/src/graphql/resolver/UserResolver.ts [hunk 7]
```diff
@@ -272,22 +276,22 @@ export class UserResolver {
...
     dbUser.lastName = lastName
     dbUser.language = language
-    dbUser.publisherId = publisherId
+    dbUser.publisherId = publisherId || 0
     dbUser.passwordEncryptionType = PasswordEncryptionType.NO_PASSWORD
-    logger.debug('new dbUser=' + dbUser)
+    logger.debug('new dbUser', dbUser)
     if (redeemCode) {
-      if (redeemCode.match(/^CL-/)) {
+      if (/^CL-/.exec(redeemCode)) {
```
Comment (Mogge) [inline backend/src/graphql/resolver/UserResolver.ts:283]: Why this? I just do not like this expressions

File: backend/src/password/EncryptorUtils.ts [hunk 8]
```diff
@@ -9,7 +12,7 @@ const sodium = require('sodium-native')
 
 // We will reuse this for changePassword
 export const isValidPassword = (password: string): boolean => {
-  return !!password.match(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^a-zA-Z0-9 \\t\\n\\r]).{8,}$/)
+  return !!/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^a-zA-Z0-9 \\t\\n\\r]).{8,}$/.exec(password)
```
Comment (Mogge) [inline backend/src/password/EncryptorUtils.ts:15]: once more

**Review by Mogge — CHANGES_REQUESTED** [review]
Please check the GDT amount carefully

File: backend/src/graphql/model/Balance.ts (hunk 1 above)
Comment (Mogge) [reply backend/src/graphql/model/Balance.ts:22]: Or float

Comment (ulfgebhardt) [reply backend/src/graphql/model/Balance.ts:22]: Float should be right! Good catch!

File: backend/src/graphql/resolver/TransactionLinkResolver.ts (hunk 5 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolver/TransactionLinkResolver.ts:131]: This is suggested by the typescript-eslint check. I am not a big fan of this - we could disable this check, but this is an equivalent

File: backend/src/graphql/model/GdtEntryList.ts (hunk 3 above)
Comment (ulfgebhardt) [reply backend/src/graphql/model/GdtEntryList.ts:31]: According to our documentation its a float!

File: backend/src/graphql/resolver/TransactionLinkResolver.ts (hunk 6 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolver/TransactionLinkResolver.ts:157]: The rule in question is this https://github.com/typescript-eslint/typescript-eslint/blob/v4.33.0/packages/eslint-plugin/docs/rules/prefer-regexp-exec.md

Money quote:
```
This rule is aimed at enforcing a consistent way to apply regular expressions to strings.
[...]
`RegExp#exec` may also be slightly faster than `String#match`; this is the reason to choose it as the preferred usage.
```

In my opinion we can disable this rule and revert related changes.

`@typescript-eslint/prefer-regexp-exec: false`

I think readability beats performance in this case - the performance is the task of the engine/compiler and not me - I have to ensure maintainability and therefore readability is king.

What you think @Mogge ?

File: backend/src/graphql/resolver/TransactionLinkResolver.ts (hunk 5 above)
Comment (ulfgebhardt) [reply backend/src/graphql/resolver/TransactionLinkResolver.ts:131]: The rule in question is this https://github.com/typescript-eslint/typescript-eslint/blob/v4.33.0/packages/eslint-plugin/docs/rules/prefer-regexp-exec.md

Money quote:
```
This rule is aimed at enforcing a consistent way to apply regular expressions to strings.
[...]
`RegExp#exec` may also be slightly faster than `String#match`; this is the reason to choose it as the preferred usage.
```

In my opinion we can disable this rule and revert related changes.

`@typescript-eslint/prefer-regexp-exec: false`

I think readability beats performance in this case - the performance is the task of the engine/compiler and not me - I have to ensure maintainability and therefore readability is king.

What you think @Mogge ?

**Review by Mogge — DISMISSED** [review]

**Review by Mogge — APPROVED** [review]

**Pull request merged**